python/mead/downloader.py

```diff
@@ -0,0 +1,300 @@
+        self.enc_dec = enc_dec
+
+    def download(self):
+        import os
+        import json
+        dload_bundle = self.dataset_desc.get("download", None)
+        if dload_bundle is not None:  # download a zip/tar/tar.gz directory, look for train, dev test files inside that.
+            try:
+                dcaches = json.load(open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json")))
+            except IOError:
+                dcaches = {}
+            if dload_bundle in dcaches and check_sanity_dir(dcaches[dload_bundle], self.dataset_desc, self.enc_dec, dload_bundle) and not self.cache_ignore:
+                download_dir = dcaches[dload_bundle]
+                print("files for {} found in cache, not downloading".format(dload_bundle))
+                return {k: os.path.join(download_dir, self.dataset_desc[k]) for k in self.dataset_desc if k.endswith("_file")}
+            else:  # try to download the bundle and unzip
+                url = dload_bundle
+                if not validate_url(url):
+                    raise RuntimeError("can not download from the given url")
+                else:
+                    cache_dir = self.data_download_cache
+                    temp_file = web_downloader(url)
+                    zipd = {'application/gzip': extract_gzip, 'application/zip': extract_zip}
+                    download_dir = extractor(filepath=temp_file, cache_dir=cache_dir, extractor_func=zipd.get(mime_type(temp_file), None))
+                    if "sha1" in self.dataset_desc:
+                        if os.path.split(download_dir)[-1] != self.dataset_desc["sha1"]:
+                            raise RuntimeError("The sha1 of the downloaded file does not match with the provided one")
+                    dcaches.update({url: download_dir})
+                    json.dump(dcaches, open(os.path.join(self.data_download_cache, "datasets-embeddings-cache.json"), "w"), indent=True)
+                    return {k: os.path.join(download_dir, self.dataset_desc[k]) for k in self.dataset_desc if k.endswith("_file")}
+        else:  # we have download links to every file or they exist
+            if not self.enc_dec:
+                return {k: SingleFileDownloader(self.dataset_desc[k], self.data_download_cache).download() for k in self.dataset_desc if k.endswith("_file")}
+            else:
+                return {k: self.dataset_desc[k] for k in self.dataset_desc if k.endswith("_file")} # these files can not be downloaded because there's a post processing on them.
+
+
+class EmbeddingDownloader(Downloader):
+    def __init__(self, embedding_file, embedding_dsz, data_download_cache, cache_ignore=False):
+        super(EmbeddingDownloader, self).__init__(data_download_cache, cache_ignore)
+        self.embedding_file = embedding_file
+        self.embedding_key = embedding_dsz
+        self.data_download_cache = data_download_cache
+
+    def _get_embedding_file(self, loc, key, dload_url):
```

Comment:
what is this supposed to return? L 259 returns a file location, L 261 returns a new copy of this class.

Reply:
if this method should return an instance of EmbeddingDownloader, then it should be marked as a class method.